tests/farmer_harvester/test_farmer.py

```diff
@@ -619,6 +638,274 @@ def assert_pool_errors_24h() -> None:
+    assert_stats_24h("stale_partials_24h")
+    assert_stats_since_start("missing_partials_since_start")
+    assert_stats_24h("missing_partials_24h")
+
+
+class DummyPoolResponse:
+    ok: bool
+    status: int
+    error_code: Optional[int]
+    error_message: Optional[str]
+    new_difficulty: Optional[int]
+
+    def __init__(
+        self,
+        ok: bool,
+        status: int,
+        error_code: Optional[int] = None,
+        error_msg: Optional[str] = None,
+        new_difficulty: Optional[int] = None,
+    ):
+        self.ok = ok
+        self.status = status
+        self.error_code = error_code
+        self.error_message = error_msg
+        self.new_difficulty = new_difficulty
+
+    async def text(self) -> str:
+        json_dict: Dict[str, Any] = dict()
+        if self.error_code:
+            json_dict["error_code"] = self.error_code
+            json_dict["error_message"] = self.error_message if self.error_message else "error-msg"
+        elif self.new_difficulty:
+            json_dict["new_difficulty"] = self.new_difficulty
+
+        return json.dumps(json_dict)
+
+    async def __aenter__(self) -> DummyPoolResponse:
+        return self
+
+    async def __aexit__(
+        self,
+        exc_type: Optional[Type[BaseException]],
+        exc_val: Optional[BaseException],
+        exc_tb: Optional[TracebackType],
+    ) -> None:
+        pass
+
+
+def create_valid_pos(farmer: Farmer) -> Tuple[farmer_protocol.NewSignagePoint, ProofOfSpace, NewProofOfSpace]:
+    case = NewProofOfSpaceCase.create_verified_quality_case(
+        difficulty=uint64(1),
+        sub_slot_iters=uint64(1000000000000),
+        pool_url="https://192.168.0.256",
+        pool_difficulty=uint64(1),
+        authentication_token_timeout=uint8(10),
+        use_invalid_peer_response=False,
+        has_valid_authentication_keys=True,
+        expected_pool_stats={},
+    )
+    sp = farmer_protocol.NewSignagePoint(
+        challenge_hash=case.challenge_hash,
+        challenge_chain_sp=case.sp_hash,
+        reward_chain_sp=std_hash(b"1"),
+        difficulty=case.difficulty,
+        sub_slot_iters=case.sub_slot_iters,
+        signage_point_index=case.signage_point_index,
+        peak_height=uint32(1),
+    )
+    pos = ProofOfSpace(
+        challenge=case.plot_challenge,
+        pool_public_key=case.pool_public_key,
+        pool_contract_puzzle_hash=case.pool_contract_puzzle_hash,
+        plot_public_key=case.plot_public_key,
+        size=case.plot_size,
+        proof=case.proof,
+    )
+    new_pos = NewProofOfSpace(
+        challenge_hash=case.challenge_hash,
+        sp_hash=case.sp_hash,
+        plot_identifier=case.plot_identifier,
+        proof=pos,
+        signage_point_index=case.signage_point_index,
+    )
+    p2_singleton_puzzle_hash = case.pool_contract_puzzle_hash
+    farmer.constants = dataclasses.replace(DEFAULT_CONSTANTS, POOL_SUB_SLOT_ITERS=case.sub_slot_iters)
```

Comment:
Maybe start from `test_constants` instead of `DEFAULT_CONSTANTS` for consistency?

Reply:
`DEFAULT_CONSTANTS` is used in another test case in `test_farmer.py`.
I tried to replace it with `test_constants` but raised an error.
Since mixing `DEFAULT_CONSTANTS` and `test_constants` in the same file also breaks consistency, I think I should leave `DEFAULT_CONSTANTS` here as it is.

Reply:
I see. We would eventually replace both of these with test constants (figuring out what led to the error you had) but that's outside the scope of this PR.